File: lua/problem_16.lua
-- OrderLog records the last N order IDs using a circular buffer.
-- record and getLast both run in O(1) time and O(N) space.

local OrderLog = {}
OrderLog.__index = OrderLog

local function newOrderLog(n)
    local buf = {}
    for i = 1, n do buf[i] = 0 end
    return setmetatable({ buf = buf, n = n, pos = 1 }, OrderLog)
end

function OrderLog:record(orderId)
    self.buf[self.pos] = orderId
    self.pos = (self.pos % self.n) + 1
end

-- getLast returns the i-th last element (i=1 is most recent).
function OrderLog:getLast(i)
    local idx = ((self.pos - 1 - i) % self.n) + 1
    return self.buf[idx]
end

local function problem16()
    print("Problem 16:")
    local log = newOrderLog(3)
    log:record(1); log:record(2); log:record(3)
    assert(log:getLast(1) == 3, "assertion failed")
    assert(log:getLast(2) == 2, "assertion failed")
    assert(log:getLast(3) == 1, "assertion failed")
    log:record(4)
    assert(log:getLast(1) == 4, "assertion failed")
    assert(log:getLast(2) == 3, "assertion failed")
    print("All Tests Passed")
end

return { problem16 = problem16 }

File: python/problem_16.py
class OrderLog:
    # OrderLog records the last N order IDs using a circular buffer.
    # record and get_last both run in O(1) time and O(N) space.
    def __init__(self, n: int):
        self._buf = [0] * n
        self._n = n
        self._pos = 0

    def record(self, order_id: int) -> None:
        # adds the order_id to the log
        self._buf[self._pos] = order_id
        self._pos = (self._pos + 1) % self._n

    def get_last(self, i: int) -> int:
        # returns the i-th last element (i=1 is most recent)
        return self._buf[(self._pos - i) % self._n]


def main():
    print("Problem 16:")
    log = OrderLog(3)
    log.record(1)
    log.record(2)
    log.record(3)
    assert log.get_last(1) == 3, "assertion failed"
    assert log.get_last(2) == 2, "assertion failed"
    assert log.get_last(3) == 1, "assertion failed"
    log.record(4)
    assert log.get_last(1) == 4, "assertion failed"
    assert log.get_last(2) == 3, "assertion failed"
    print("All Tests Passed")
